### ISSUE TITLE
Field gets repeated with default settings on save of a section

### ISSUE THREAD

**Issue by pixelninja** [symphonycms/symphony-2#2736]
Affected Symphony version(s) : 2.7.0
PHP version(s) : 5.6.31
OS(es) : MacOS

This is a really strange one, which I've come across twice now so I thought I should log it incase I'm not the only one. Only problem is I can't replicate it or find a pattern to it.

Issue: When creating or editing a section, on save a field is duplicated. I've experienced it with both the Expression Field, and the Textbox Field which is why I'm posting the bug here instead of with the extension. 

The original field is saved with all the parameters fine, the duplicated field has the same label assigned to it but default field parameters. 

Then when I try to edit the duplicated field, it doesn't retain any of the changes. If I remove it, it remains on save until you refresh the browser.

Like I said, it's a weird one which I can't find a pattern to for testing :(

Anyone else experienced this? I would bet on it being an issue with the duplicator.

**Comment by pixelninja** [comment]
I found the pattern! Create a textbox or expression field (just seems to be these 2 for me). Save the section, then try and change the Placement field. It will create a second field, instead of updating the original field. If you try and remove either field, it won't remove it. And every time the section successfully saves, a new field is added. The only way around it is to remove all fields of that type, save the section, and re-add it.

**Comment by nitriques** [comment]
@pixelninja check out https://github.com/symphonists/expressionfield/commit/adb1a95c46e274ac2d9f9aeedd89eabdc4bc3f23 if you want to see the fix. I'll leave open for now, but this will be close as soon as 2.7.1 is out. Thanks!

**Comment by pixelninja** [comment]
Awesome work @nitriques, it fixes it.

For anyone else having this issue, I did have to remove the existing field and any duplicates completely. The re-added field won't duplicate.

**Comment by nitriques** [comment]
@pixelninja I really feel like the fix belongs in extensions. I've patch a couple of other one. I'll tag as Won't fix for now, just for triage.

**Comment by twiro** [comment]
I noticed the same behaviour with **[number field](https://github.com/symphonycms/numberfield)** (latest version) after an update to Symphony 2.7.0.

I tried the fix @nitriques proposed for **expression field** and that first didn't help.

Only after I manually copied all number field instances from `sym_fields` to `sym_fields_number` (which was empty before) the fix started to have an impact and avoid duplication of fields when saving sections.

So to properly fix the affected extensions we'd also need to include an update script that checks the database for missing data and fills in the missing records while updating I guess…

**Comment by twiro** [comment]
mmh... look like this already has been fixed for number field a while ago:

https://github.com/symphonycms/numberfield/issues/13

But as the fields in my project were created before the numberfield got a commit function and as no update-script refilled the records my `sym_fields_number` was always empty… and I didn't notice until updating to Symphony 2.7.0.

So I guess there's nothing really to fix in number field – the only thing helpful would be a script that checks the records in`sym_fields_number` each time the extension is activated/updated. But that might be too much work for a rare edge case like this… just documenting it here in case anyone else is stumbling across these problems.

**Comment by twiro** [comment]
As the extension **[member last visit](https://github.com/twiro/member_last_visit)** was also affected from the **"field duplication issue"** I took the chance and had a closer look into what exactly is going on while working on an update for the extension.

### General summary of the issue

- Field extensions without a `commit()`-function do not populate their own `sym_fields_XYZ`-tables
- This obviously had no negative side-effects until Symphony 2.7 as the basic field settings in `sym_fields` was sufficient to make everything work as expected.
- In Symphony 2.7.0 field without an entry in their corresponding `sym_fields_XYZ`-table get duplicated when editing a section.
- The original field will still remain without a proper entry in the `sym_fields_XYZ`-table after the duplication and therefore will cause another duplication with each saving of the section.
- The only way to stop this is to either delete the original field and replace it with a new instance (which means loosing all data of the field) or manually filling in the missing entries in the `sym_fields_XYZ`-table

### So far, so clear – but now to the more difficult aspect of the problem:

The missing `commit()`-function in **[Number Field](https://github.com/symphonycms/numberfield)** has been [added three years ago](https://github.com/symphonycms/numberfield/commit/b382f9716b6d9eff777f9da19394c1949767d038) in version 1.7 but I was still running into the "field duplication issue" when updating a project to Symphony 2.7 **because I did not resave my sections in the period between updating to Number Field 1.7 and updating to Symphony 2.7** – only when would the subsequently added `commit()`-function have been sufficient for avoiding field duplication, because the missing entries in `sym_fields_number` would have been silently re-added without any errors or side-effects.

But if you update to Symphony 2.7 and entries in `sym_fields_XYZ`-table are still missing you're screwed and the `commit()`-function alone won't save you.

### Extension based solution

While working on the update for the **member last visit** extension I included an update script that would also handle the above-mentioned scenario by [repopulating the `sym_fields_member_last_visit`-table](https://github.com/twiro/member_last_visit/blob/master/extension.driver.php#L136-L156) with data from `sym_fields` automatically when the updater runs.

This solves the problem and I would suggest to include something similar in all affected extensions if this bug can not be easily avoided in Symphony Core.

### Core based solution

But if it would easily be possible to fix this in the core I'd vote for going that way and including a general solution in Symphony 2.7.1 as this field-duplication-bug is highly annoying and might cause lots of irritation in the long run.

As far as I see all affected extensions don't rely on any custom settings and therefore come with a `sym_fields_XYZ`-table that only consists of `id` and `field_id`. Wouldn't it be possible to simply include a solution in the core that is prepared for the scenario that entries might be missing in that table when saving a section and that is able to do the job my custom [`populateFieldTable()`](https://github.com/twiro/member_last_visit/blob/master/extension.driver.php#L136-L156)-function does for **member last visit** for any extension that doesn't need any custom settings/data in their `sym_fields_XYZ`-table?

> _I've added that check in 2.7.0 to prevent bigger, nastier error, but did not anticipate this bug._

I'm not familiar with everything that's going on in the core when saving a section but I assume something like this should be perfectly possible without loosing the prevention of the bigger, nastier errors you mention here @nitriques

**Comment by michael-e** [comment]
@twiro thanks a lot for your research! I agree that tis behaviour is annoying, and my gut feeling is that this will hit a lot of people badly.

According to your description, even if extensions have been fixed, this issue may occur for fields which have been added before the extension was fixed. How bad.

Why the hell does the core react this way if a field has no custom settings? If a field that is provided by an extension simply uses `parent::commit`, things should not break. To me this looks like a bug in the core.

@nitriques why do you say that this will not be fixed in the core? Is it too complicated? And why does this suddenly happen with Sym 2.7?

**Comment by nitriques** [comment]
Thanks @twiro and @michael-e 

I did not knew that this fix would create such confusion. The problem lies in fact with the fact that some extensions did not create settings entries for their field, while the core was expecting it. This led to a problem where we would try to update non existing records and could lead to locking a table that did not exists. From what I could find at the time, it was a good idea to make sure that fields would always create both records, in tbl_fields and tbl_fields_XYZ.

I'll try to find a fix for this. I'll try to do it first in the updater, but may end up patching in directly in EntryManager, which is something I would try to avoid.

The real fix would be to make the commit function abstract, we would be sure that it is implemented. I'll see.

**Comment by nitriques** [comment]
I've found a core based solution: #2787 please comment !
